ddos_drivers/src/vga_text.rs

```diff
@@ -12,10 +12,69 @@ pub struct Writer {
 }
 
 impl Writer {
+    pub fn handle_escapes(&mut self, ch: u8) -> bool {
+        // Handles escape characters for writing to the screen.
+        //
+        // Done: \n, \t
+        // TODO: All the other escape characters (\r, etc...)
+        //
+        // Returns true if the character is an escape character, false if it's not.
+        // If it's an escape character, then the `x` and `y` positions do not
+        // have to be incremented because that's handled inside this function.
+        match ch {
+            b'\n' => {
+                self.y += 1;
+                self.x = 0;
+                return true;
+            }
+            b'\t' => {
+                self.write_string("    ");
+                return true;
+            }
+            _ => {
+                return false;
+            }
+        }
+    }
+
+    pub fn handle_scrolling(&mut self) {
+        // Handles scrolling
+        // In essence this just pushes everything up by decrementing their `y` position by 1
+        // This overwrites row 0
+        if self.y < HEIGHT {
+            return;
+        }
+        for i in WIDTH..(WIDTH * HEIGHT) {
+            self.buf[i - WIDTH] = self.buf[i];
```

Comment:
Probably more efficient to use `core::ptr::volatile_copy_nonoverlapping_memory` to copy rows, and in my experience speed is worthwhile here

Reply:
you forgot to remove this